nbtlib/path.py

```diff
@@ -0,0 +1,265 @@
+            new_accessors = (ListIndex(index=None),)
+        elif isinstance(key, Compound):
+            new_accessors = (CompoundMatch(key),)
+        else:
+            raise KeyError(key)
+
+        accessors = tuple(self)
+
+        for accessor in new_accessors:
+            accessors = extend_accessors(accessors, accessor)
+
+        return self.from_accessors(accessors)
+
+    def __add__(self, other):
+        if isinstance(other, Path):
+            return self[other]
+        elif isinstance(other, str):
+            return self[Path(other)]
+        else:
+            return NotImplemented
+
+    def __radd__(self, other):
+        if isinstance(other, Path):
+            return other[self]
+        elif isinstance(other, str):
+            return Path(other)[self]
+        else:
+            return NotImplemented
+
+    def __eq__(self, other):
+        if isinstance(other, str):
+            other = Path(other)
+        return super().__eq__(other)
+
+    def __ne__(self, other):
+        if isinstance(other, str):
+            other = Path(other)
+        return super().__ne__(other)
+
+    def __hash__(self):
+        return super().__hash__()
+
+    @classmethod
+    def from_accessors(cls, accessors=()):
+        return super().__new__(cls, accessors)
+
+    def traverse(self, tag):
+        tags = [(None, tag)]
+
+        setter = None
+        deleter = None
+
+        for accessor in self:
+            setter = getattr(accessor, 'set', setter)
+            deleter = getattr(accessor, 'delete', deleter)
+
+            tags = accessor.get(tags)
+
+        return tags, setter, deleter
+
+    def get(self, tag):
+        return [tag for _, tag in self.traverse(tag)[0]]
+
+    def set(self, tag, value):
+        tags, setter, _ = self.traverse(tag)
+
+        if setter:
+            setter(tags, value)
+
+    def delete(self, tag):
+        tags, _, deleter = self.traverse(tag)
+
+        if deleter:
+            deleter(tags)
+
+    def __repr__(self):
+        return f'{self.__class__.__name__}({str(self)!r})'
+
+    def __str__(self):
+        segments = ['']
+
+        for accessor in self:
+            segment = str(accessor)
+
+            if not segment or segment.startswith('['):
+                segments[-1] += segment
+
+            elif segment.startswith('{'):
+                if segments[-1].endswith('[]'):
+                    segments[-1] = segments[-1][:-2] + f'[{segment}]'
+                else:
+                    segments[-1] += segment
+
+            else:
+                segments.append(segment)
+
+        return '.'.join(filter(None, segments))
+
+
+def parse_accessors(path):
+    try:
+        parser = Parser(tokenize(path))
+    except InvalidLiteral:
+        return ()
+
+    while True:
+        try:
+            tag = parser.parse()
+        except InvalidLiteral as exc:
+            raise InvalidPath(f'Invalid path at position {exc.args[0][0]}') from exc
+
+        if isinstance(tag, String):
```

Comment:
Yeah I did it on purpose, the constructor is more forgiving than minecraft's parser. I even have [tests](https://github.com/vberlier/nbtlib/blob/nbt-paths/tests/test_path.py#L164) for that. However, no matter how the path was constructed, calling `str()` will always result in a string compatible with minecraft. If you use `''` they will turn into into `""`.

```python
>>> str(Path("'hello \"world\"'"))
'"hello \\"world\\""'
```

Also yeah right now I quote everything that doesn't match [this regex](https://github.com/vberlier/nbtlib/blob/nbt-paths/nbtlib/path.py#L212) because even though minecraft could handle the unquoted version, I think it's a bit more readable when you start sprinkling weird characters.

```python
>>> str(Path('this-.-looks--.-weird-'))
'"this-"."-looks--"."-weird-"'
```

Not sure if it makes a big difference though, I can change it if you want.

Reply:
```
>>> Path("a'b'")
Path('a.b')
```

```
>>> Path(".2b")
Path('2b')
>>> Path(".2d")
Traceback (most recent call last):
nbtlib.path.InvalidPath: Invalid path element 0.2d
```
🙃

Reply:
Hmm... yeah that's a problem, I think I have an easy fix though :thinking:

Reply:
Paths with numbers as keys are now working properly d6255c7 but I think it's going to be a bit more difficult for the one with single quotes...